compiler/rustc_const_eval/src/interpret/operand.rs

```diff
@@ -246,6 +246,57 @@ impl<'tcx, Tag: Provenance> ImmTy<'tcx, Tag> {
     }
 }
 
+/// See `try_read_immediate_from_mplace` for what this function is about.
+/// Exposed to the rest of the crate since some hacky code in `eval_queries.rs` (that shouldn't
+/// exist) also needs it.
+pub(crate) fn type_is_scalar<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> bool {
+    match ty.kind() {
+        // Primitive types. We *have to* handle these since the primitive operations that
+        // act on them require it.
+        ty::Bool
+        | ty::Char
+        | ty::Float(_)
+        | ty::Int(_)
+        | ty::Uint(_)
+        | ty::RawPtr(..)
+        | ty::Ref(..)
+        | ty::FnPtr(..)
+        | ty::FnDef(..)
+        | ty::Never => true,
+        // Compound types. We have to exclude unions here.
+        ty::Adt(adt_def, substs) => {
+            if adt_def.is_union() {
+                // Unions are explicitly allowed to be partially initialized, so we have to
+                // exclude them here.
+                false
+            } else {
+                // Check all fields of all variants, to make sure there is no union hiding here.
+                adt_def.all_fields().all(|field| type_is_scalar(tcx, field.ty(tcx, substs)))
+            }
+        }
+        ty::Tuple(..) => {
+            // If all fields are scalar, we are good.
+            ty.tuple_fields().iter().all(|field| type_is_scalar(tcx, field))
+        }
+        // FIXME: can these ever have Scalar ABI?
+        ty::Closure(..) | ty::Generator(..) => false,
```

Comment:
This line is the problem: some closures (and likely generators) *can* end up with a `Scalar` layout, and they can even have niches, which then means validity will call `read_scalar` and cause an ICE when we do not treat it as a scalar.

To fix this, we need a way to iterate over the types of all the fields of (all the variants of) a closure/generator. Any suggestion for how to do that? :D I browsed the code for a bit and found nothing.

Reply:
For closures you'll want https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.ClosureSubsts.html#method.upvar_tys which you get by copying the substs into a new instance of ClosureSubsts. The same helper exists for generators. I guess we should make this less cumbersome to do

Reply:
What about renaming this to `type_can_be_scalar` and returning true in all cases except unions?

Reply:
Then we'd still have the bug for (e.g.) closures that capture unions in a way that they still get `abi::Scalar` layout.

Reply:
I am experimenting with changing the layout to support undef scalars. If perf is happy with it, we can just ask the layout whether the scalars are undef and not do an immediate representation then.